### ISSUE TITLE
Sync-related object store deadlocks

### ISSUE THREAD

**Issue by austinzheng** [realm/realm-object-store#178]
I believe these are possible deadlocks, and related issues. Making this ticket to track my progress as I investigate further:
- When a session dies, it takes the session lock while it waits for changes to upload. If, while waiting, an error occurs, it is reported via the error handler. Certain errors may require a change in state, thus the error handler attempts to acquire the lock the dying-session lambda is already holding, and the program stops. (EDIT: this applies more generally to the error handler. The contract reads as follows:)

> CAUTION: The specified callback function may be called before the call to bind() returns, and it may be called (or continue to execute) after the session object is destroyed. The application must pass a handler that can be safely called, and can safely continue to execute from the point in time where bind() starts executing, and up until the point in time where the last invocation of `clint.run()` returns. Here, `client` refers to the associated Client object.
- A couple of the session handle APIs (`isValid`, `logOut`) seem to sporadically deadlock the program on acquiring the state change lock.
- (Not a deadlock) If a session is dying (waiting for pending changes to upload) and the user logs out, there should be a way to kill the session immediately. Right now, the session is only killed immediately if it is active (or not otherwise dying).
- (Not a deadlock) The object store `Session` should only run the `wait_for_download` (or its upload equivalent) APIs if the state makes sense to do so.
- (Not a deadlock) `testImmediateSessionTerminationWhenLoggingOut` sometimes causes a `bad_server_file_ident` error during the cleanup phase, which uncovered one of the deadlocks above.

**Comment by mrackwitz** [comment]
I'm seeing a concrete dead-lock triggered by a change in the Node.js SDK as it seems. It can reproduced by running the tests, specifically `testCurrent`,  on https://github.com/realm/realm-js/pull/679.

**Comment by austinzheng** [comment]
Is there a way to tell what line of the test is causing the deadlock? `console.log()` is unhelpful. Do we have good reason to believe this is an object store deadlock, based on the codepath which that test is exercising?

**Comment by austinzheng** [comment]
Never mind, after talking to @alazier and @bdash we have a pretty good idea as to what's going on.

**Comment by bdash** [comment]
@mrackwitz, the deadlock I see when testing with your realm-js patch looks identical to what #219 fixed.

**Comment by alazier** [comment]
@bdash can you rebase the js branch on master? Don't think we have those changes yet. I just tried but am unfamiliar with the added functionality and don't have time to dig in right now. Can test it out once the changes are in though.

**Comment by alazier** [comment]
I just tried using object-store master and the deadlock still persists, so I'm not sure that commit resolves the issue.

**Comment by bdash** [comment]
Ah, the deadlock manifests in the same way as the one #219 fixed, but that's only because whenever the call to `SyncUser::register_session` within `SyncManager::get_session` throws, we'll deadlock. I'll fix the deadlock, but the real issue @mrackwitz will need to have addressed is that an exception is being thrown:
```
-> 181 	        throw std::invalid_argument("Can only register sessions that haven't previously been registered.");
```

**Comment by bdash** [comment]
The deadlock is fixed by #242, and I've filed #243 about the limitation that is resulting in the exception being thrown.
